core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1499,29 +1501,104 @@ class KafkaApis(val requestChannel: RequestChannel,
   def handleLeaveGroupRequest(request: RequestChannel.Request) {
     val leaveGroupRequest = request.body[LeaveGroupRequest]
 
-    // the callback for sending a leave-group response
-    def sendResponseCallback(error: Errors) {
-      def createResponse(requestThrottleMs: Int): AbstractResponse = {
-        val response = new LeaveGroupResponse(new LeaveGroupResponseData()
+    val isRequestAuthorized = !authorize(request.session, Read, Resource(Group, leaveGroupRequest.data.groupId, LITERAL))
+    // Before version 3, leave group request is still in single mode
+    if (leaveGroupRequest.version <= 2) {
+      handleSingleLeaveGroup(request,
```

Comment:
I feel we should be able to consolidate these paths. A single leave group is just a batch of size one. As far as I can tell, the only difference is that a batch leave group of size 1 would have a top-level error code. But that seems like something that could be dealt with when building the response.